src/resmom/linux/alps.c

```diff
@@ -5815,17 +5815,63 @@ alps_confirm_reservation(job *pjob)
 int
 alps_cancel_reservation(job *pjob)
 {
-	char buf[1024];
-	basil_response_t *brp;
+	char			buf[1024];
+	basil_response_t	*brp;
+	char			filename[MAXPATHLEN];
 
 	if (!pjob) {
 		log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_NODE, LOG_NOTICE, __func__,
 			"Cannot cancel ALPS reservation, invalid job.");
 		return (-1);
 	}
+
+	/*
+	 * Get the name of the file that contains ALPS reservation ID
+	 * information.  We may need to read the file, or we just need to
+	 * unlink it.  Either way, we need the filename.
+	 */
+	snprintf(filename, MAXPATHLEN, "%s/aux/%s.alpsresv", pbs_conf.pbs_home_path, pjob->ji_qs.ji_jobid);
+	
+	if (pjob->ji_extended.ji_ext.ji_reservation < 0) {
+		       /*
+			* The job structure doesn't have ALPS reservation ID information.
+			* This can happen when we hit various race conditions.  To help
+			* with this issue, we have previously saved the ALPS resv ID
+			* in a file.
+			*
+			* The format of the file will consist of one line entry that
+			* has the ALPS reservation ID number. We retrieve it here so
+			* the ALPS reservation can be canceled.
+			*/
+		FILE    *fp = NULL;
+		long    value;
+
+		fp = fopen(filename, "r");
+		if (fp == NULL) {
+			snprintf(log_buffer, sizeof(log_buffer),
+				"Unable to open the file %s containing the ALPS Reservation ID", filename);
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+				(char *)__func__, log_buffer);
+			return (-1);
+		}
+		if (fscanf(fp, "%ld", &value) != 1) {
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+				(char *)__func__, "failed to read the ALPS reservation ID value");
+			(void)fclose(fp);
+			(void)unlink(filename);
+			return (-1);
+		}
+		(void)fclose(fp);
+		pjob->ji_extended.ji_ext.ji_reservation = value;
+	}
+
+	       /* At this point we have no further use for the file that contains
+		* the ALPS reservation ID.  Get rid of the file.
+		*/
```

Comment:
too verbose again. We can remove this comment block. If you decide to summarize, please align it with the code block below.